worker/provisioner/provisioner_task.go

```diff
@@ -689,101 +723,362 @@ func (task *provisionerTask) maintainMachines(machines []*apiprovisioner.Machine
+	} else {
+		sort.Sort(ByPopulationThenName(task.availabilityZoneMachines))
+		for _, zoneMachines := range task.availabilityZoneMachines {
+			if !zoneMachines.FailedMachineIds.Contains(machine.Id()) {
+				machineZone = zoneMachines.ZoneName
+				zoneMachines.MachineIds.Add(machine.Id())
+				break
+			}
 		}
+	}
+	return machineZone
+}
 
-		startInstanceParams, err := constructStartInstanceParams(
-			task.controllerUUID,
-			m,
-			instanceCfg,
-			pInfo,
-			possibleTools,
-		)
-		if err != nil {
-			return task.setErrorStatus("cannot construct params for machine %q: %v", m, err)
+type ByPopulationThenName []*AvailabilityZoneMachine
+
+func (b ByPopulationThenName) Len() int {
+	return len(b)
+}
+
+func (b ByPopulationThenName) Less(i, j int) bool {
+	switch {
+	case b[i].MachineIds.Size() < b[j].MachineIds.Size():
+		return true
+	case b[i].MachineIds.Size() == b[j].MachineIds.Size():
+		return b[i].ZoneName < b[j].ZoneName
+	}
+	return false
+}
+
+func (b ByPopulationThenName) Swap(i, j int) {
+	b[i], b[j] = b[j], b[i]
+}
+
+// startMachines starts a goroutine for each specified machine to
+// start it.  The specified machines are provided the name of the
+// first availability zone to try.  Errors from individual start
+// machine attempts will be logged.
+func (task *provisionerTask) startMachines(machines []*apiprovisioner.Machine) error {
+	if len(machines) == 0 {
+		return nil
+	}
+
+	// Get the distributionGroups for each machine now to avoid
+	// successive calls to DistributionGroupByMachineId which will
+	// return the same data.
+	machineTags := make([]names.MachineTag, len(machines))
+	for i, machine := range machines {
+		machineTags[i] = machine.MachineTag()
+	}
+	machineDistributionGroups, err := task.distributionGroupFinder.DistributionGroupByMachineId(machineTags...)
+	if err != nil {
+		return err
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(machines))
+	errMachines := make([]error, len(machines))
+
+	for i, m := range machines {
+		// Make sure we shouldn't be stopping before we start the next machine
+		select {
+		case <-task.catacomb.Dying():
+			return task.catacomb.ErrDying()
+		default:
 		}
 
-		if err := task.startMachine(m, pInfo, startInstanceParams); err != nil {
-			return errors.Annotatef(err, "cannot start machine %v", m)
+		if machineDistributionGroups[i].Err != nil {
+			return err
+		}
+
+		go func(machine *apiprovisioner.Machine, dg []string, index int) {
+			defer wg.Done()
+			pInfo, startInstanceParams, err := task.setupToStartMachine(machine)
+			if err != nil {
+				errMachines[index] = err
+				return
+			}
+			if err := task.startMachine(machine, pInfo, startInstanceParams, dg); err != nil {
+				task.removeMachineFromAZMap(machine)
+				errMachines[index] = err
+			}
+		}(m, machineDistributionGroups[i].MachineIds, i)
+	}
+
+	// TODO ProvisionerParallelization 2017-10-09
+	// Is this necessary?
+	wg.Wait()
+	for _, err := range errMachines {
```

Comment:
not sure this is correct - why not return a slice of errors

Reply:
combined them into one, any of them will cause the provisioner to restart.